python_modules/dagster/dagster/_core/execution/plan/inputs.py

```diff
@@ -976,19 +979,79 @@ def __new__(
         )
 
-    @property
-    def resolved_by_step_key(self) -> str:
-        return self.unresolved_step_output_handle.resolved_by_step_key
+    def get_resolving_handle(self):
+        return self.unresolved_step_output_handle.resolution_source.get_resolving_handle()
 
-    @property
-    def resolved_by_output_name(self) -> str:
-        return self.unresolved_step_output_handle.resolved_by_output_name
+    # @property
+    # def resolved_by_step_key(self) -> str:
+    #     return self.unresolved_step_output_handle.resolved_by_step_key
+
+    # @property
+    # def resolved_by_output_name(self) -> str:
+    #     return self.unresolved_step_output_handle.resolved_by_output_name
 
-    def resolve(self, mapping_key: str) -> FromStepOutput:
+    def resolve(self, mapping_key: str) -> Union[FromStepOutput, FromPendingDynamicStepOutput]:
         check.str_param(mapping_key, "mapping_key")
-        return FromStepOutput(
+        output_handle = self.unresolved_step_output_handle.resolve(mapping_key)
+        if isinstance(output_handle, StepOutputHandle):
+            return FromStepOutput(
+                step_output_handle=output_handle,
+                fan_in=False,
+            )
+        else:
+            return FromUnresolvedStepOutput(
+                unresolved_step_output_handle=output_handle,
+            )
+
+    def get_step_output_handle_dep_with_placeholder(self) -> StepOutputHandle:
+        return self.unresolved_step_output_handle.get_step_output_handle_with_placeholder()
+
+    def required_resource_keys(self, _pipeline_def: PipelineDefinition) -> Set[str]:
+        return set()
+
+
+# prototype note: feels bad to have to add another but this models it cleanly
+@whitelist_for_serdes
+class FromUnresolvedDynamicStepOutput(
+    NamedTuple(
+        "_FromUnresolvedDynamicStepOutput",
+        [
+            ("unresolved_step_output_handle", UnresolvedStepOutputHandle),
+        ],
+    ),
+):
+    """
+    This step input source models being downstream of unresolved step that has a dynamic output.
+    """
+
+    def __new__(
+        cls,
+        unresolved_step_output_handle: UnresolvedStepOutputHandle,
+    ):
+        return super().__new__(
+            cls,
+            unresolved_step_output_handle=check.inst_param(
+                unresolved_step_output_handle,
+                "unresolved_step_output_handle",
+                UnresolvedStepOutputHandle,
+            ),
+        )
+
+    # @property
+    # def resolved_by_step_key(self) -> str:
+    #     return self.unresolved_step_output_handle.resolved_by_step_key
+
+    # @property
+    # def resolved_by_output_name(self) -> str:
+    #     return self.unresolved_step_output_handle.resolved_by_output_name
+    def get_resolving_handle(self):
+        return self.unresolved_step_output_handle.resolution_source.get_resolving_handle()
+
+    def resolve(self, mapping_key: str) -> FromPendingDynamicStepOutput:
```

Comment:
[2] ^ the proper test case should reveal that this needs to behave more like the one above